src/Microsoft.Health.Dicom.Functions/BlobMigration/CleanupDeletedDurableFunction.Orchestration.cs

```diff
@@ -0,0 +1,110 @@
+        EnsureArg.IsNotNull(input, nameof(input));
+        EnsureArg.IsNotNull(input.Batching, nameof(input.Batching));
+
+        logger = context.CreateReplaySafeLogger(logger);
+
+        // Get the max watermark of all deleted instances for the exit scenario
+        long maxWatermark = input.MaxWatermark ?? await context.CallActivityWithRetryAsync<long>(
+                 nameof(GetMaxDeletedChangeFeedWatermarkAsync),
+                 _options.RetryOptions,
+                 new CleanupDeletedBatchArguments
+                 {
+                     FilterTimeStamp = input.FilterTimeStamp
+                 });
+
+        IReadOnlyList<ChangeFeedEntry> changeFeedEntries = await context.CallActivityWithRetryAsync<IReadOnlyList<ChangeFeedEntry>>(
+            nameof(GetDeletedChangeFeedInstanceBatchesAsync),
+            _options.RetryOptions,
+            new CleanupDeletedBatchArguments
+            {
+                StartWatermark = input.Completed.HasValue ? input.Completed.Value.Start : default,
+                EndWatermark = input.Completed.HasValue ? input.Completed.Value.End : default,
+                BatchSize = input.Batching.Size,
+                FilterTimeStamp = input.FilterTimeStamp == DateTime.MinValue ? null : input.FilterTimeStamp,
+            });
+
+        bool hasReachedToEnd = input.Completed.HasValue && input.Completed.Value.End >= maxWatermark;
+
+        if (changeFeedEntries.Count > 0)
+        {
+            var batchRange = new WatermarkRange(changeFeedEntries[0].Sequence, changeFeedEntries[^1].Sequence);
+
+            logger.LogInformation("Beginning to delete the range {Range}.", batchRange);
+
+            var instanceIdentifiers = changeFeedEntries.Select(x => new VersionedInstanceIdentifier(x.StudyInstanceUid, x.SeriesInstanceUid, x.SopInstanceUid, x.OriginalVersion)).ToList();
+            await context.CallActivityWithRetryAsync(nameof(CleanupDeletedBatchAsync), _options.RetryOptions, instanceIdentifiers);
+
+            // Create a new orchestration with the same instance ID to process the remaining data
+            logger.LogInformation("Completed deleting the range {Range}. Total files deleted in range: '{NumFiles}'. " +
+                "Continuing with new execution...", batchRange, batchRange.End - batchRange.Start + 1);
+
+            WatermarkRange nextRange = new WatermarkRange(batchRange.End + 1, batchRange.End + input.Batching.Size);
+
+            // Using completed to populate next set of ranges
+            context.ContinueAsNew(
+                new BlobMigrationCheckpoint
+                {
+                    Completed = nextRange,
+                    CreatedTime = input.CreatedTime ?? await context.GetCreatedTimeAsync(_options.RetryOptions),
+                    Batching = input.Batching,
+                    MaxWatermark = maxWatermark,
+                });
+        }
+        else if (!hasReachedToEnd && input.Completed.HasValue)
```

Comment:
```suggestion
        else if (!input.Completed.HasValue || input.Completed.Value.End < maxWatermark)
```

Reply:
I see!

Reply:
I will say that the behavior for the sproc is very confusing haha. The use of a time filter on the first request also implies there should always be at least 1 value is returned if there is anything to delete, whereas the watermarks are well defined. In fact, you probably didn't need to track the current "end" and only needed the start. As in, "I have deleted up to watermark 42, now please give me the next 50?

Furthermore, are the change feed watermarks here semantically different from the instance watermarks? Are these sequences unique to the change feed such that they're actually sequential?

Reply:
We should probably have some commits detailing these nuances while the code exists

Reply:
Yes, I agree we could have used only one sequence for this logic.  And yes changefeed watermark are sequentially and unique. I would have separated the sproc like one using TimeStamp and other using Watermark, but reduced since I am going to delete it once this is executed.